gems/smithy-schema/lib/smithy-schema/document/serializer.rb

```diff
@@ -0,0 +1,283 @@
+          elsif value == -::Float::INFINITY
+            '-Infinity'
+          elsif value.to_f.nan?
+            'NaN'
+          else
+            value.to_f
+          end
+        end
+
+        def list(ref, values, opts)
+          values.collect do |value|
+            next if value.nil?
+
+            shape(ref.shape.member, value, opts)
+          end
+        end
+
+        def map(ref, values, opts)
+          values.each.with_object({}) do |(key, value), data|
+            next if value.nil?
+
+            data[key.to_s] = shape(ref.shape.value, value, opts)
+          end
+        end
+
+        def structure(ref, values, opts)
+          values.each_pair.with_object({}) do |(k, v), h|
+            next if v.nil?
+
+            if (member_ref = resolve_member_ref(ref, k))
+              member_name = resolve_member_name(member_ref, opts)
+              h[member_name] = shape(member_ref, v, opts)
+            end
+          end
+        end
+
+        def timestamp(ref, value, opts)
+          value = normalize_timestamp_value(value)
+          return value.to_i unless opts[:use_timestamp_format]
+
+          trait = 'smithy.api#timestampFormat'
+          case ref.traits[trait] || ref.shape.traits[trait]
+          when 'date-time' then value.utc.iso8601
+          when 'http-date' then value.utc.httpdate
+          else
+            # default to epoch-seconds
+            value.to_i
+          end
+        end
+
+        def union(ref, values, opts)
+          data = {}
+          if values.is_a?(Smithy::Schema::Union)
+            member_ref = ref.shape.member_by_type(values.class)
+            member_name = resolve_member_name(member_ref, opts)
+            data[member_name] = shape(member_ref, values, opts)
+          else
+            key, value = values.first
+            if (member_ref = resolve_member_ref(ref, key))
+              member_name = resolve_member_name(member_ref, opts)
+              data[member_name] = shape(member_ref, value, opts)
+            end
+          end
+          data
+        end
+
+        def format_document_data(shape, data, opts = {})
+          ref = ShapeRef.new(shape: shape)
+          document_data = shape(ref, data, opts)
+          document_data['__type'] = shape.id
+          document_data
+        end
+
+        def discriminator?(data)
+          data.is_a?(Hash) && data.key?('__type')
+        end
+
+        def normalize_timestamp_value(value)
+          case value
+          when Time then value
+          when Numeric then Time.at(value)
+          else Time.parse(value)
+          end
+        end
+
+        def resolve_member_ref(ref, name)
+          ref.shape.member(name) || find_member_ref_by_names(ref, name)
+        end
+
+        def find_member_ref_by_names(ref, name)
+          ref.shape.members.values.find do |r|
+            r.traits['smithy.api#jsonName'] == name || r.member_name == name
+          end
+        end
+
+        def resolve_member_name(member_ref, opts)
```

Comment:
Check out the location_name approach in my PR - we should use similar terms. You can easily handle this with || optionality.